spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/pubsub/brave/TracingMessageReceiver.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.trace.pubsub.brave;
+
+import brave.Span;
+import brave.Tracer;
+import brave.propagation.TraceContextOrSamplingFlags;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.pubsub.v1.PubsubMessage;
+
+import static brave.Span.Kind.CONSUMER;
+
+final class TracingMessageReceiver implements MessageReceiver {
+	private final PubSubTracing pubSubTracing;
+
+	private final MessageReceiver delegate;
+
+	private final String subscriptionName;
+
+	TracingMessageReceiver(MessageReceiver delegate, PubSubTracing pubSubTracing, String subscriptionName) {
+		this.pubSubTracing = pubSubTracing;
+		this.delegate = delegate;
+		this.subscriptionName = subscriptionName;
+	}
+
+	@Override
+	public void receiveMessage(PubsubMessage pubsubMessage, AckReplyConsumer ackReplyConsumer) {
+		// instrument message
+		PubsubMessage.Builder messageBuilder = pubsubMessage.toBuilder();
+		PubSubConsumerRequest request = new PubSubConsumerRequest(messageBuilder, subscriptionName);
+		TraceContextOrSamplingFlags extracted =
+				pubSubTracing.extractAndClearTraceIdHeaders(pubSubTracing.consumerExtractor, request, messageBuilder);
+
+		Span consumerSpan = pubSubTracing.nextMessagingSpan(pubSubTracing.consumerSampler, request, extracted);
+		Span listenerSpan = pubSubTracing.tracer.newChild(consumerSpan.context());
+
+		if (!consumerSpan.isNoop()) {
+			consumerSpan.name("next-message").kind(CONSUMER); // TODO: do we need this consumer span at all?
+			consumerSpan.tag("pubsub.subscription", subscriptionName); // TODO: shouldn't have to tag manually since it's in PubSubConsumerRequest
+			if (pubSubTracing.remoteServiceName != null) {
+				consumerSpan.remoteServiceName(pubSubTracing.remoteServiceName);
+			}
+
+			// incur timestamp overhead only once
+			long timestamp = pubSubTracing.tracing.clock(consumerSpan.context()).currentTimeMicroseconds();
+			consumerSpan.start(timestamp);
+			long consumerFinish = timestamp + 1L; // save a clock reading
```

Comment:
I am confused by why start and finish are basically the same. Should the start of `consumerSpan` be before we run the user callback `delegate.receiveMessage(...)` and the end after?

This seems to be what `listenerSpan` does, but then why do we need `consumerSpan` at all?

Update: haha, played myself by answering your TODO question. I still want to know what the visual looks like with and without `consumerSpan`, but IMHO we don't need the fake-ish `consumerSpan`.

Update2: found an example of what this looks like in the sample (great sample, btw). Yeah, "next-message" is completely redundant.

Reply:
Again this is following the Brave examples. So, I'm a bit worried about breaking consistency with those instrumentations.